pyo3-macros-backend/src/introspection.rs

```diff
@@ -50,52 +52,123 @@ pub fn module_introspection_code<'a>(
-            ("id", IntrospectionNode::IntrospectionId(None)),
+            ("id", IntrospectionNode::IntrospectionId(Some(ident))),
             ("name", IntrospectionNode::String(name)),
+            ("arguments", arguments_introspection_data(signature)),
         ]
         .into(),
     )
-    .emit(pyo3_crate_path);
-    let introspection_id = introspection_id_const();
-    quote! {
-        #stub
-        #introspection_id
+    .emit(pyo3_crate_path)
+}
+
+fn arguments_introspection_data<'a>(signature: &'a FunctionSignature<'a>) -> IntrospectionNode<'a> {
+    let mut argument_desc = signature.arguments.iter().filter_map(|arg| {
+        if let FnArg::Regular(arg) = arg {
+            Some(arg)
+        } else {
+            None
+        }
+    });
+
+    let mut arguments = Vec::new();
+
+    for (i, param) in signature
+        .python_signature
+        .positional_parameters
+        .iter()
+        .enumerate()
+    {
+        let arg_desc = if let Some(arg_desc) = argument_desc.next() {
+            arg_desc
+        } else {
+            panic!("Less arguments than in python signature");
+        };
+        arguments.push(argument_introspection_data(
+            param,
+            if i < signature.python_signature.positional_only_parameters {
+                "POSITIONAL_ONLY"
+            } else {
+                "POSITIONAL_OR_KEYWORD"
+            },
+            arg_desc,
+        ));
+    }
+
+    if let Some(param) = &signature.python_signature.varargs {
+        arguments.push(IntrospectionNode::Map(
+            [
+                ("name", IntrospectionNode::String(param)),
+                ("kind", IntrospectionNode::String("VAR_POSITIONAL")),
+            ]
+            .into(),
+        ));
+    }
+
+    for (param, _) in &signature.python_signature.keyword_only_parameters {
+        let arg_desc = if let Some(arg_desc) = argument_desc.next() {
+            arg_desc
+        } else {
+            panic!("Less arguments than in python signature");
+        };
+        arguments.push(argument_introspection_data(param, "KEYWORD_ONLY", arg_desc));
+    }
+
+    if let Some(param) = &signature.python_signature.kwargs {
+        arguments.push(IntrospectionNode::Map(
+            [
+                ("name", IntrospectionNode::String(param)),
+                ("kind", IntrospectionNode::String("VAR_KEYWORD")),
+            ]
+            .into(),
+        ));
+    }
+
+    IntrospectionNode::List(arguments)
+}
+
+fn argument_introspection_data<'a>(
+    name: &'a str,
+    kind: &'a str,
+    desc: &'a RegularArg<'_>,
+) -> IntrospectionNode<'a> {
+    let mut params: HashMap<_, _> = [
+        ("name", IntrospectionNode::String(name)),
+        ("kind", IntrospectionNode::String(kind)),
+    ]
+    .into();
+    if desc.default_value.is_some() {
+        // TODO: generate a nice default values for literals (None, false, 0, ""...)
+        params.insert("default_value", IntrospectionNode::String("..."));
```

Comment:
We should re-use the machinery we have for text signature docstrings, see `default_value_for_parameter()` in `signature.rs`.

Reply:
Thank you! I missed it. Done.